File: tests/st/ops/test_ops_select_ext.py
# Copyright 2024 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
import pytest
import numpy as np
import mindspore as ms
from mindspore import Tensor, mint, jit
from tests.st.ops.test_tools.test_op import TEST_OP
from tests.st.utils import test_utils
from tests.mark_utils import arg_mark


@test_utils.run_with_cell
def select_ext_forward_func(x, dim, index):
    return mint.select(x, dim, index)


@test_utils.run_with_cell
def select_ext_backward_func(x, dim, index):
    return ms.grad(select_ext_forward_func)(x, dim, index)

def GenInputData(np_data_type, shape=(3, 4, 5)):
    """GenInputData"""
    size = 1
    for s in shape:
        size *= s
    data = np.arange(size).reshape(*shape).astype(np_data_type)
    return Tensor(data)


@arg_mark(plat_marks=['platform_ascend'], level_mark='level1',
          card_mark='onecard', essential_mark='essential')
@pytest.mark.parametrize('mode', ['pynative'])
def test_ops_select_ext(mode):
    """
    Feature: pyboost function.
    Description: test function select forward.
    Expectation: expect correct result.
    """
    input_tensor_list = [
        np.array([[2, 3, 5, 1], [3, 1, 2, 6]], dtype=np.float16)
    ]
    expect_list = [
        np.array([3, 1], dtype=np.float16)
    ]
    expect_grad_list = [
        np.array([[0, 1, 0, 0], [0, 1, 0, 0]], dtype=np.float16)
    ]
    dim_list = [1]
    index_list = [1]

    for i in range(len(input_tensor_list)):
        x = input_tensor_list[i]
        dim = dim_list[i]
        index = index_list[i]
        if mode == 'pynative':
            ms.context.set_context(mode=ms.PYNATIVE_MODE)
            output = select_ext_forward_func(ms.Tensor(x), dim, index)
            out_grad = select_ext_backward_func(ms.Tensor(x), dim, index)
        elif mode == 'KBK':
            ms.context.set_context(mode=ms.GRAPH_MODE)
            output = (jit(select_ext_forward_func, jit_level="O0"))(ms.Tensor(x), dim, index)
            out_grad = (jit(select_ext_backward_func, jit_level="O0"))(ms.Tensor(x), dim, index)
        else:
            ms.context.set_context(mode=ms.GRAPH_MODE)
            output = (jit(select_ext_forward_func, backend="GE"))(ms.Tensor(x), dim, index)
            out_grad = (jit(select_ext_backward_func, backend="GE"))(ms.Tensor(x), dim, index)
        expect = expect_list[i]
        expect_grad = expect_grad_list[i]
        np.testing.assert_allclose(output.asnumpy(), expect, rtol=1e-3)
        np.testing.assert_allclose(out_grad.asnumpy(), expect_grad, rtol=1e-3)


@arg_mark(plat_marks=['platform_ascend'], level_mark='level1',
          card_mark='onecard', essential_mark='unessential')
def test_select_ext_dynamic_shape():
    """
    Feature: Test select with dynamic shape in pynative mode.
    Description: call mint.select with valid input, dim and index.
    Expectation: return the correct value.
    """
    ms_data1 = ms.Tensor(np.array([[2, 3, 5, 1], [3, 1, 2, 6]], dtype=np.float16))
    dim1 = 0
    index1 = 0

    ms_data2 = ms.Tensor(np.array([[[2, 3, 5], [3, 1, 2]], [[2, 3, 5], [3, 1, 2]]], dtype=np.float16))
    dim2 = 1
    index2 = 1
    TEST_OP(select_ext_forward_func, [[ms_data1, dim1, index1], [ms_data2, dim2, index2]],
            disable_mode=['GRAPH_MODE_GE', 'GRAPH_MODE_O0'],
            disable_case=['EmptyTensor', 'ScalarTensor'])

File: tests/ut/cpp/ops/test_ops_select_ext.cc
/**
 * Copyright 2024 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <vector>
#include "ir/tensor_new.h"
#include "ops/test_ops_cmp_utils.h"
#include "ir/dtype/number.h"
#include "infer/ops_func_impl/select_ext_view.h"
#include "ops/test_value_utils.h"
#include "abstract/dshape.h"
#include "utils/tensor_construct_utils.h"
#include "mindspore/ops/op_def/auto_generate/gen_ops_primitive_s.h"

namespace mindspore {
namespace ops {

struct SelectExtParams {
  ShapeVector input_shape;
  TypePtr     input_dtype;
  ValuePtr    dim;
  ValuePtr    index;
  ShapeVector output_shape;
  TypePtr     output_dtype;
};

class TestSelectExt : public TestOps, public testing::WithParamInterface<SelectExtParams> {};

TEST_P(TestSelectExt, dyn_shape) {
  const auto &param = GetParam();
  auto x = std::make_shared<abstract::AbstractTensor>(param.input_dtype, param.input_shape);
  ASSERT_NE(x, nullptr);
  auto dim = param.dim->ToAbstract();
  ASSERT_NE(dim, nullptr);
  auto index = param.index->ToAbstract();
  ASSERT_NE(index, nullptr);

  auto expect_shape = std::make_shared<abstract::Shape>(param.output_shape);
  auto expect_type = std::make_shared<TensorType>(param.output_dtype);
  DoFuncImplInferAndCompare<SelectExtViewFuncImpl>(kNameSelectExtView, {x, dim, index}, expect_shape, expect_type);
}

static std::vector<SelectExtParams> GetCases() {
  auto dyn_rank = abstract::TensorShape::kShapeRankAny;
  auto dyn_dim = abstract::TensorShape::kShapeDimAny;
  std::vector<SelectExtParams> cases = {
    SelectExtParams{{4, 2, 3}, kFloat16, CreateScalar<int64_t>(1), CreateScalar<int64_t>(1), {4, 3}, kFloat16},
    SelectExtParams{{dyn_rank}, kFloat16, CreateScalar<int64_t>(1), CreateScalar<int64_t>(1), {dyn_rank}, kFloat16},
    SelectExtParams{{4, 2, 3}, kFloat16, CreateScalar(kValueAny), CreateScalar<int64_t>(1), {dyn_rank}, kFloat16},
    SelectExtParams{{4, dyn_dim, 3}, kFloat16, CreateScalar<int64_t>(1), CreateScalar<int64_t>(1), {4, 3}, kFloat16},
  };
  return cases;
}

class TestSelectExtSimple : public TestOps, public testing::WithParamInterface<SelectExtParams> {};

TEST_P(TestSelectExtSimple, simple_infer) {
  const auto &param = GetParam();
  auto x = tensor::from_spec(param.input_dtype->type_id(), param.input_shape, device::DeviceType::kCPU);
  auto dim = param.dim->ToAbstract();
  auto index = param.index->ToAbstract();

  auto expect_shape = std::make_shared<abstract::Shape>(param.output_shape);
  auto expect_type = std::make_shared<TensorType>(param.output_dtype);

  DoFuncImplInferAndCompare<SelectExtViewFuncImpl>(kNameSelectExtView, {x->ToAbstract(), dim, index}, expect_shape, expect_type);
}

INSTANTIATE_TEST_CASE_P(TestSelectExt, TestSelectExt, testing::ValuesIn(GetCases()));

INSTANTIATE_TEST_CASE_P(
  TestSelectExtSimple, TestSelectExtSimple,
  testing::Values(
    SelectExtParams{{4, 2, 3}, kFloat16, CreateScalar<int64_t>(1), CreateScalar<int64_t>(1), {4, 3}, kFloat16},
    SelectExtParams{{4, 2, 3}, kFloat16, CreateScalar<int64_t>(0), CreateScalar<int64_t>(1), {2, 3}, kFloat16}
));
}  // namespace ops
}  // namespace mindspore
